authorizer-app-backend/src/main/java/org/radarbase/authorizer/resources/RestSourceUserResource.kt

```diff
@@ -61,32 +61,41 @@ class RestSourceUserResource(
     fun query(
         @QueryParam("project-id") projectId: String?,
         @QueryParam("source-type") sourceType: String?,
+        @QueryParam("search") search: String?,
         @DefaultValue(Integer.MAX_VALUE.toString()) @QueryParam("size") pageSize: Int,
         @DefaultValue("1") @QueryParam("page") pageNumber: Int,
     ): RestSourceUsers {
+        projectId ?: throw HttpBadRequestException(
+            "missing_project_id",
+            "Project ID parameter is mandatory",
+        )
+
         val sanitizedSourceType = if (sourceType != null && sourceType in sourceClientService) {
             sourceType
         } else null
 
-        val projects = if (projectId != null) {
-            auth.checkPermissionOnProject(Permission.SUBJECT_READ, projectId)
-            listOf(projectId)
-        } else {
-            projectService.userProjects(auth, Permission.SUBJECT_READ)
-                .map { it.id }
-        }
+        auth.checkPermissionOnProject(Permission.SUBJECT_READ, projectId)
 
-        if (projects.isEmpty()) return RestSourceUsers(emptyList(), Page(totalElements = 0))
+        val minimalSearch = search?.takeIf { it.length >= 2 }
 
-        val queryPage = Page(pageNumber = pageNumber, pageSize = pageSize)
-        val (records, page) = userRepository.query(queryPage, projects, sanitizedSourceType)
+        val userIds = if (minimalSearch != null) {
+            projectService.projectUsers(projectId)
+                .mapNotNull { sub ->
+                    val externalId = sub.externalId ?: return@mapNotNull null
+                    sub.id.takeIf { minimalSearch in externalId }
```

Comment:
The external user id in the database is the user ID from Fitbit/Garmin, not the MP external user ID, which I suppose is what users would search for. Since that info is not in the database, it’s searched before running SQL. It’s run before to ensure that the paging works correctly. I’ve imposed a minimum character length of 2 to reduce the number of matches.

Reply:
Ah ok, I see. Sorry got confused between the terminology of`externalUserId` and `serviceUserId`